File: Diagrammatize.Design.Pattern/code/cxx/AbstractFactory/A2/listfactory/ListTray.cpp
#include "ListTray.hpp"
#include <sstream>

ListTray::ListTray(const std::string& caption): Tray(caption) {
}

std::string ListTray::makeHTML() {
    std::ostringstream buffer;
    buffer << "<li>\n";
    buffer << caption << "\n";
    buffer << "<ul>\n";
    for (auto item : tray) {
        buffer << item->makeHTML();
    }
    buffer << "</ul>\n";
    buffer << "</li>\n";
    return buffer.str();
}

File: Diagrammatize.Design.Pattern/code/python/AbstractFactory/Sample/listfactory/ListTray.py
import io
from factory import Tray

class ListTray(Tray):
    def __init__(self, caption):
        super().__init__(caption)

    def makeHTML(self):
        buffer = io.StringIO()
        buffer.write("<li>\n")
        buffer.write(self.caption + "\n")
        buffer.write("<ul>\n")
        for item in self.tray:
            buffer.write(item.makeHTML())
        buffer.write("</ul>\n")
        buffer.write("</li>\n")
        return buffer.getvalue()
